From c8a4ad27e3075b74e1861df6feaf7078655a909d Mon Sep 17 00:00:00 2001
From: Sathyanarayanan Kuppuswamy <sathyanarayanan.kuppuswamy@intel.com>
Date: Fri, 14 Dec 2012 15:26:13 -0800
Subject: [PATCH 65/78] WIFI[PATCH 2/2]: Move gpio resource allocation from
 platform code to wifi driver

BZ: 75109

Currently we are allocating wifi gpio resources in platform code. This
patches move it to driver.

Change-Id: Id4a6f88b7d501bf7d551b6d51054abaae0f32fe8
Signed-off-by: Sathyanarayanan Kuppuswamy <sathyanarayanan.kuppuswamy@intel.com>
Signed-off-by: Aymen Zayet <aymen.zayet@intel.com>
Reviewed-on: http://android.intel.com:8080/81871
Reviewed-by: Yang, Fei <fei.yang@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/sdio.c |   13 +++++++++++++
 wl12xx-compat/include/linux/wl12xx.h             |    5 +++++
 2 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/sdio.c b/wl12xx-compat/drivers/net/wireless/wl12xx/sdio.c
index 10565b8..d78b943 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/sdio.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/sdio.c
@@ -267,6 +267,12 @@ static int __devinit wl1271_probe(struct sdio_func *func,
 		goto out_free_glue;
 	}
 
+	if (wlan_data->hw_init) {
+		ret = wlan_data->hw_init(wlan_data);
+		if (ret)
+			goto out_free_glue;
+	}
+
 	/* if sdio can keep power while host is suspended, enable wow */
 	mmcflags = sdio_get_host_pm_caps(func);
 	dev_dbg(glue->dev, "sdio PM caps = 0x%x\n", mmcflags);
@@ -329,12 +335,19 @@ out:
 static void __devexit wl1271_remove(struct sdio_func *func)
 {
 	struct wl12xx_sdio_glue *glue = sdio_get_drvdata(func);
+	struct wl12xx_platform_data *wlan_data;
 
+	wlan_data = wl12xx_get_platform_data();
+	BUG_ON(!wlan_data);
 	/* Undo decrement done above in wl1271_probe */
 	pm_runtime_get_noresume(&func->dev);
 
 	platform_device_del(glue->core);
 	platform_device_put(glue->core);
+	/* Free platform resources */
+	if (wlan_data->hw_deinit)
+		wlan_data->hw_deinit(wlan_data);
+
 	kfree(glue);
 
 }
diff --git a/wl12xx-compat/include/linux/wl12xx.h b/wl12xx-compat/include/linux/wl12xx.h
index 0d63731..4719d23 100644
--- a/wl12xx-compat/include/linux/wl12xx.h
+++ b/wl12xx-compat/include/linux/wl12xx.h
@@ -48,8 +48,13 @@ enum {
 
 struct wl12xx_platform_data {
 	void (*set_power)(bool enable);
+	int (*hw_init)(struct wl12xx_platform_data *pdata);
+	void (*hw_deinit)(struct wl12xx_platform_data *pdata);
 	/* SDIO only: IRQ number if WLAN_IRQ line is used, 0 for SDIO IRQs */
 	int irq;
+	/* gpio must be set to -EINVAL by platform code if gpio based irq is
+	not used */
+	int gpio;
 	bool use_eeprom;
 	int board_ref_clock;
 	int board_tcxo_clock;
-- 
1.7.9

